src/nbl/video/utilities/CComputeBlit.cpp

```diff
@@ -0,0 +1,152 @@
+		static_cast<uint32_t>(inImageType) + 1u,
+		alphaBinCount);
+
+	auto gpuUnspecShader = device->createGPUShader(std::move(cpuShaderOverriden));
+
+	return device->createGPUSpecializedShader(gpuUnspecShader.get(), { nullptr, nullptr, "main" });
+}
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> CComputeBlit::createNormalizationSpecializedShader(const asset::IImage::E_TYPE inImageType, const asset::E_FORMAT outImageFormat,
+	const asset::E_FORMAT outImageViewFormat, const core::vectorSIMDu32& workgroupDims, const uint32_t alphaBinCount)
+{
+	char shaderStart[] = "#version 460 core\n#include \"nbl/builtin/glsl/blit/default_compute_normalization.comp\"\n";
+	auto shader = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(strlen(shaderStart), shaderStart, core::adopt_memory);
+
+	auto cpuShader = core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(shader), asset::ICPUShader::buffer_contains_glsl, asset::IShader::ESS_COMPUTE, "CComputeBlit::createNormalizationSpecializedShader");
+
+	const char* outImageFormatGLSLString = getGLSLFormatStringFromFormat(outImageFormat);
+	const char* glslFormatQualifier = getGLSLFormatStringFromFormat(outImageViewFormat);
+
+	// Todo(achal): Remove
+	char formatInclude[1024] = "";
+	if (outImageFormat != outImageViewFormat)
+		snprintf(formatInclude, sizeof(formatInclude), "#include <nbl/builtin/glsl/blit/blit/formats/%s.glsl>\n", outImageFormatGLSLString);
+
+	auto cpuShaderOverriden = asset::IGLSLCompiler::createOverridenCopy(cpuShader.get(),
+		"#define _NBL_GLSL_WORKGROUP_SIZE_X_ %d\n"
+		"#define _NBL_GLSL_WORKGROUP_SIZE_Y_ %d\n"
+		"#define _NBL_GLSL_WORKGROUP_SIZE_Z_ %d\n"
+		"#define _NBL_GLSL_BLIT_DIM_COUNT_ %d\n"
+		"#define _NBL_GLSL_BLIT_ALPHA_BIN_COUNT_ %d\n"
+		"#define _NBL_GLSL_BLIT_NORMALIZATION_OUT_IMAGE_FORMAT_ %s\n"
+		"%s\n" // _NBL_GLSL_BLIT_NORMALIZATION_SOFTWARE_CODEC_
+		"%s\n", // Todo(achal): Remove format include in favour of the symbol gotten from outImageFormatGLSLString
+		inImageType >= asset::IImage::ET_1D ? workgroupDims.x : 1u,
+		inImageType >= asset::IImage::ET_2D ? workgroupDims.y : 1u,
+		inImageType >= asset::IImage::ET_3D ? workgroupDims.z : 1u,
+		static_cast<uint32_t>(inImageType) + 1u,
+		alphaBinCount,
+		glslFormatQualifier,
+		outImageFormat != outImageViewFormat ? "#define _NBL_GLSL_BLIT_NORMALIZATION_SOFTWARE_CODEC_\n" : "",
+		reinterpret_cast<const char*>(formatInclude));
+
+	auto gpuUnspecShader = device->createGPUShader(std::move(cpuShaderOverriden));
+
+	return device->createGPUSpecializedShader(gpuUnspecShader.get(), { nullptr, nullptr, "main" });
+}
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> CComputeBlit::createBlitSpecializedShader(const asset::E_FORMAT inImageFormat, const asset::E_FORMAT outImageFormat,
+	const asset::E_FORMAT outImageViewFormat, const asset::IImage::E_TYPE inImageType, const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent,
```

Comment:
again, only want `inImageViewFormat` and `outImageViewFormat`

btw the image types kinda have to match so maybe just name the type argument `imageType`

Reply:
I use `outImageFormat` to determine the count of shared memory floats given its size, like:
```C++
const uint32_t smemFloatCount = m_availableSharedMemory/(sizeof(float)*outChannelCount);
shaderSourceStream << "#define _NBL_GLSL_BLIT_SMEM_FLOAT_COUNT_ " << smemFloatCount << "\n";
```

Can I get rid of this?

Reply:
### Bad argument names

again, my point is, I only care about the **view's** format (the view that the user intends to use, not any additional views blit needs to create to be able to SW encode), not the image format.

There really shouldn't be a single instance of `...ImageFormat`, just `..ImageViewFormat`s

### Do we need to know the formats

Due to using HW filtered sampling, you dont need to know input format, only type.

You can work out what image view format needs to be used for the storage image, from the format the user wants, internally... no need for both parameters.

I guess if your code fully aliases shared memory now and decides what range is used for what using push contstants, you don't need to care about number of channels, just as the user (as an argument) for the limit on smem usage.

Reply:
Resolved.